### PR TITLE
fix: oauth v2 response handling

### PR DESCRIPTION
**Pull request by sanpj2292**
# Description

### Issue

The OAuth interceptor currently alters the response status code in specific scenarios reserved for OAuth destinations:
- Validation failure before `http.RoundTrip`: Sets a 500 status code with a validation error response.
- Validation failure/unmarshal failure after `http.RoundTrip`: Sets a 500 status code with the respective failure's error response.
- Successful refresh token action: Sets a 500 status code with the response from the transformer.
- Refresh token failure due to network errors: Sets a 500 status code with the response from the OAuth interceptor (including connection errors).
- Success/failure of `authStatusInactive`: Sets a 400 status code with the response from either `rudder-auth` or the transformer.
- Any erroneous response from the transformer: Retains the status code from the transformer and sends the response from the transformer.

Directly altering HTTP responses and status codes is causing issues in `RouterTransform`, resulting in consistent 500 errors for any of the above interception failures. Even when the transformer fails with a 4xx status code, we encounter 500 errors in `jobsDB` through this new client, which is not ideal.

`RouterTransform` expects a 200 status code consistently, and any deviation from this creates problems for us. To address this, we've modified the implementation to capture both the response from the transformer (`OriginalResponse`) and from the interceptor (`InterceptorResponse`).

### Implemented Solution

We now always set `{"originalResponse": "response from transformer", "interceptorResponse": {"statusCode": , "response": "response from interceptor" }}` to the responseBody when errors occur during the refreshToken/authStatusInactive flows.

We need to unmarshal the response and determine which response to send to downstream modules(router, regulation-worker).

Additionally, we've included a couple of tests to verify the new implementation in this PR.

## Linear Ticket

Resolves INT-956

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: router/transformer/transformer.go [hunk 1]
```diff
@@ -357,6 +376,22 @@ func (trans *handle) ProxyRequest(ctx context.Context, proxyReqParams *ProxyRequ
 		}
 	}
 
+	/*
+		respData will be in ProxyResponseV0 or ProxyResponseV1
+	*/
+	transportResponse := oauthv2.TransportResponse{} // response that we get from oauth-interceptor in postRoundTrip
+	if (*trans.oauthV2EnabledLoader).Load() {
+		_ = json.Unmarshal(respData, &transportResponse)
```
Comment (koladilip) [inline router/transformer/transformer.go:384]: what about err handling

File: router/worker.go [hunk 2]
```diff
@@ -780,36 +782,35 @@ func (w *worker) proxyRequest(ctx context.Context, destinationJob types.Destinat
...
-			trRespBody:     proxyRequestResponse.ProxyRequestResponseBody,
-			secret:         m[0].Secret,
-			contentType:    proxyRequestResponse.RespContentType,
-		}, proxyRequestResponse.OAuthErrorCategory)
-
-	} else {
-		respStatusCode = proxyRequestResponse.ProxyRequestStatusCode
-		respBodyTemp = proxyRequestResponse.ProxyRequestResponseBody
-		contentType = proxyRequestResponse.RespContentType
+	if routerutils.IsNotEmptyString(string(authType)) && authType == oauth.OAuth {
```
Comment (koladilip) [inline router/worker.go:793]: authType == oauth.OAuth is not sufficient?

File: router/transformer/transformer.go (hunk 1 above)
Comment (sanpj2292) [reply router/transformer/transformer.go:384]: No need to handle it as upstream you can get a string response even before performing OAuth operations like Refresh Token, Auth Status Toggle. It's alright if the respData structure doesn't follow oauthv2.TransportResponse struct

File: regulation-worker/internal/delete/api/api.go [hunk 3]
```diff
@@ -142,14 +143,36 @@ func (api *APIManager) deleteWithRetry(ctx context.Context, job model.Job, desti
 	if err != nil {
 		return model.JobStatus{Status: model.JobStatusFailed, Error: err}
 	}
+	stCd := resp.StatusCode
+	respBodyBytes := bodyBytes
+	// Post response work to be done for OAuthV2
+	if isOAuth && api.IsOAuthV2Enabled {
+		transportResponse := oauthv2.TransportResponse{} // initing to prevent panic
+		err = json.Unmarshal(bodyBytes, &transportResponse)
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply regulation-worker/internal/delete/api/api.go:151]: No need to handle it as upstream you can get a string response even before performing OAuth operations like Refresh Token, Auth Status Toggle. It's alright if the respData structure doesn't follow oauthv2.TransportResponse struct

File: router/transformer/transformer.go [hunk 4]
```diff
@@ -205,6 +207,15 @@ func (trans *handle) Transform(transformType string, transformMessage *types.Tra
 		trans.logger.Errorf("[Router Transfomrer] :: Transformer returned status code: %v reason: %v", resp.StatusCode, resp.Status)
 	}
 
+	transResp := oauthv2.TransportResponse{}
+	if (*trans.oauthV2EnabledLoader).Load() {
+		// Enabled only for OAuth destinations that are using V2
+		err = json.Unmarshal(respData, &transResp)
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/transformer/transformer.go:213]: No need to handle it as upstream you can get a string response even before performing OAuth operations like Refresh Token, Auth Status Toggle. It's alright if the respData structure doesn't follow oauthv2.TransportResponse struct

File: router/transformer/transformer.go [hunk 5]
```diff
@@ -230,6 +241,14 @@ func (trans *handle) Transform(transformType string, transformMessage *types.Tra
 		var out []int64
 		invalid := make(map[int64]struct{}) // invalid jobIDs are the ones that are in the response but were not included in the request
 		for i := range destinationJobs {
+			isValidOAuthCategory := oauthv2.IsValidAuthErrorCategory(destinationJobs[i].AuthErrorCategory)
+			if isValidOAuthCategory && transResp.InterceptorResponse.StatusCode > 0 {
+				destinationJobs[i].StatusCode = transResp.InterceptorResponse.StatusCode
+			}
+			if isValidOAuthCategory && strings.TrimSpace(transResp.InterceptorResponse.Response) != "" {
+				// Should this be set to `error` alone ?
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/transformer/transformer.go]: Yes.

File: router/worker.go (hunk 2 above)
Comment (sanpj2292) [reply router/worker.go]: Sufficient

File: router/transformer/transformer.go [hunk 6]
```diff
@@ -230,6 +241,14 @@ func (trans *handle) Transform(transformType string, transformMessage *types.Tra
 		var out []int64
 		invalid := make(map[int64]struct{}) // invalid jobIDs are the ones that are in the response but were not included in the request
 		for i := range destinationJobs {
+			isValidOAuthCategory := oauthv2.IsValidAuthErrorCategory(destinationJobs[i].AuthErrorCategory)
+			if isValidOAuthCategory && transResp.InterceptorResponse.StatusCode > 0 {
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply router/transformer/transformer.go]: Done

File: regulation-worker/internal/delete/api/api.go [hunk 7]
```diff
@@ -142,14 +143,36 @@ func (api *APIManager) deleteWithRetry(ctx context.Context, job model.Job, desti
 	if err != nil {
 		return model.JobStatus{Status: model.JobStatusFailed, Error: err}
 	}
+	stCd := resp.StatusCode
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply regulation-worker/internal/delete/api/api.go]: Done

**Review by koladilip — APPROVED** [review]

**Pull request merged**